File: ClaudeRelayAndroid/core-protocol/src/main/kotlin/relay/protocol/TokenGenerator.kt
package relay.protocol

import java.security.MessageDigest
import java.security.SecureRandom
import java.util.Base64

/**
 * Generates cryptographically secure API tokens and provides hashing/validation
 * utilities.
 *
 * Ports `TokenGenerator.swift`: 32 secure-random bytes → Base64URL (no padding,
 * 43 chars) plaintext, plus its lowercase-hex SHA-256 digest (64 chars). The
 * Swift type also builds a `TokenInfo`; here [generate] returns the
 * `(plaintext, hash)` pair the Android client needs.
 */
object TokenGenerator {
    private val secureRandom = SecureRandom()
    private val urlEncoder = Base64.getUrlEncoder().withoutPadding()

    /**
     * Generates a new random token.
     *
     * @return `Pair(plaintext, hash)` — the 43-char Base64URL plaintext and its
     *   64-char lowercase-hex SHA-256 digest.
     */
    fun generate(): Pair<String, String> {
        val bytes = ByteArray(32)
        secureRandom.nextBytes(bytes)
        val plaintext = urlEncoder.encodeToString(bytes)
        return plaintext to hash(plaintext)
    }

    /** Computes the lowercase-hex SHA-256 digest of [token]. */
    fun hash(token: String): String {
        val digest = MessageDigest.getInstance("SHA-256").digest(token.toByteArray(Charsets.UTF_8))
        return digest.joinToString("") { "%02x".format(it) }
    }

    /** Validates [token] against a previously stored SHA-256 hex [storedHash]. */
    fun validate(token: String, storedHash: String): Boolean = hash(token) == storedHash
}

File: Sources/ClaudeRelayKit/Security/TokenGenerator.swift
import Foundation
#if canImport(CryptoKit)
import CryptoKit
#else
import Crypto
#endif

// MARK: - Data Extension

extension Data {
    /// Base64URL encoding without padding (RFC 4648 Section 5).
    /// Single-pass transformation instead of 3 separate string scans.
    func base64URLEncodedString() -> String {
        var result = base64EncodedString()
        result.unicodeScalars.removeAll(where: { $0 == "=" })
        return String(result.map { ch in
            switch ch {
            case "+": return "-"
            case "/": return "_"
            default: return ch
            }
        })
    }
}

// MARK: - TokenGenerator

/// Generates cryptographically secure API tokens and provides hashing/validation utilities.
public enum TokenGenerator {

    /// Generates a new random token and its associated `TokenInfo`.
    ///
    /// - Parameters:
    ///   - label: Optional human-readable label for the token.
    ///   - expiryDays: Number of days until the token expires. `nil` means never expires.
    /// - Returns: A tuple of the plaintext token (43 chars, base64URL) and its `TokenInfo`.
    public static func generate(label: String? = nil, expiryDays: Int? = nil) -> (plaintext: String, info: TokenInfo) {
        // 32 cryptographically secure random bytes
        let bytes = SecureRandom.bytes(count: 32)

        // Base64URL encode (no padding) -> 43 characters
        let plaintext = Data(bytes).base64URLEncodedString()

        // SHA-256 hash of the plaintext
        let tokenHash = hash(plaintext)

        // Short ID from UUID
        let id = String(UUID().uuidString.prefix(8)).lowercased()

        let now = Date()
        let expiresAt = expiryDays.map { now.addingTimeInterval(Double($0) * 86400) }

        let info = TokenInfo(
            id: id,
            tokenHash: tokenHash,
            label: label,
            createdAt: now,
            expiresAt: expiresAt
        )

        return (plaintext, info)
    }

    /// Computes the SHA-256 hex digest of a token string.
    ///
    /// - Parameter token: The plaintext token.
    /// - Returns: Lowercase hex-encoded SHA-256 hash.
    public static func hash(_ token: String) -> String {
        let digest = SHA256.hash(data: Data(token.utf8))
        return digest.map { String(format: "%02x", $0) }.joined()
    }

    /// Validates a plaintext token against a stored hash.
    ///
    /// - Parameters:
    ///   - token: The plaintext token to validate.
    ///   - storedHash: The previously stored SHA-256 hex hash.
    /// - Returns: `true` if the token matches the stored hash.
    public static func validate(_ token: String, against storedHash: String) -> Bool {
        hash(token) == storedHash
    }
}
